### DISCUSSION TITLE
CMYK to sRGB profiles fail if I use image operations before reading the input file

### DISCUSSION THREAD

**Discussion by alanorth** [ImageMagick/ImageMagick#6234] [Help]
### ImageMagick version

7.1.1.5-1

### Operating system, version and so on

Arch Linux

### Description

Did something change in ImageMagick's syntax/behavior, or this is a bug? With ImageMagick 7.1.1.5 on Arch Linux, creating an sRGB thumbnail from a [CMYK PDF](https://hdl.handle.net/10568/116598) distorts the colors if I apply [image operations](https://legacy.imagemagick.org/script/command-line-processing.php#option) like `-define` or `-density` before reading the input file:

<p align="center">
  <img src="https://user-images.githubusercontent.com/191754/230285170-be56108c-d8ee-4d80-a7f0-c4567bba3870.jpg" height="300" />
</p>

The command used was:

```console
$ convert -define pdf:use-cropbox=true \
    10568-116598.pdf\[0\] \
    -profile /usr/share/ghostscript/iccprofiles/default_cmyk.icc \
    -profile /usr/share/ghostscript/iccprofiles/srgb.icc \
    /tmp/im7.jpg
```

With ImageMagick 6.9.11.60 on Ubuntu 22.04 it works as expected:

<p align="center">
  <img src="https://user-images.githubusercontent.com/191754/230285868-4e67c9fd-18a4-4819-b9b3-e569cd11b9b5.jpg" height="300" />
</p>

The command used was:

```console
$ convert -define pdf:use-cropbox=true \
    10568-116598.pdf\[0\] \
    -profile /usr/share/ghostscript/9.55.0/iccprofiles/default_cmyk.icc \
    -profile /usr/share/ghostscript/9.55.0/iccprofiles/srgb.icc \
    /tmp/im6.jpg
```

If I omit the `-define` on ImageMagick 7.1.1.5 the colors are correct, but the CropBox is of course messed up (see the erroneous two-page layout):

<p align="center">
  <img src="https://user-images.githubusercontent.com/191754/230286176-6b80df6d-2804-48bb-89f8-d3c5d90e70ad.jpg" height="300" />
</p>

Thank you!

**Comment by fmw42** [comment]
To convert a PDF/PS/EPS file from CMYK to sRGB, you should use -colorspace sRGB before reading the vector file.

**Comment by snibgo** [comment]
As fmw42 says. I also suggest using `magick` so you get the current v7 syntax instead of the old and inferior v6 syntax.
```
magick -colorspace sRGB -define pdf:use-cropbox=true Ethopia.pdf[0] x.jpg
```

**Reply by alanorth** [reply]
Thanks @fmw42 and @snibgo. The color is off with this one though too:

```console
$ magick -colorspace sRGB -define pdf:use-cropbox=true Ethopia.pdf[0] x.jpg
```

See the pink band under the maroon "project report" band which doesn't exist in the PDF or the other thumbnails above:

<p align="center">
  <img src="https://user-images.githubusercontent.com/191754/230750983-b6af29bb-e8e5-4e6c-a1d8-aa6bb3afae7c.jpg" width=300" />
</p>

Pardon the naive question, but regarding the `-colorspace` without profiles, does that mean ImageMagick will approximate the CMYK→sRGB conversion? In this case it seems that particular PDF has an embedded profile, but I don't know how common that is. Does ImageMagick use it?

> _I also suggest using magick so you get the current v7 syntax instead of the old and inferior v6 syntax_

Oh, I didn't know that. Thanks.

**Comment by fmw42** [comment]
That first page of the PDF does not have an embedded color profile according to Imagemagick identify -verbose


```
image:
  Filename: PR_Nutrition_Ethiopia_PK.pdf
  Base filename: PR_Nutrition_Ethiopia_PK.pdf
  Permissions: rw-r--r--
  Format: PDF (Portable Document Format)
  Mime type: application/pdf
  Class: DirectClass
  Geometry: 1191x842+0+0
  Resolution: 72x72
  Print size: 16.5417x11.6944
  Units: Undefined
  Colorspace: CMYK
  Type: ColorSeparation
  Endianness: Undefined
  Depth: 16/8-bit
  Channel depth:
    cyan: 8-bit
    magenta: 8-bit
    yellow: 8-bit
    black: 8-bit
  Channel statistics:
    Pixels: 1002822
    Cyan:
      min: 0  (0)
      max: 61937 (0.945098)
      mean: 6621.72 (0.101041)
      standard deviation: 12076.9 (0.184282)
      kurtosis: 2.89467
      skewness: 1.88559
      entropy: 0.253528
    Magenta:
      min: 0  (0)
      max: 64507 (0.984314)
      mean: 6059.68 (0.0924647)
      standard deviation: 12944.4 (0.19752)
      kurtosis: 6.12759
      skewness: 2.54908
      entropy: 0.215343
    Yellow:
      min: 0  (0)
      max: 65535 (1)
      mean: 6693.29 (0.102133)
      standard deviation: 13557.7 (0.206877)
      kurtosis: 5.03833
      skewness: 2.3836
      entropy: 0.249864
    Black:
      min: 0  (0)
      max: 58082 (0.886275)
      mean: 2768.5 (0.0422447)
      standard deviation: 10695.5 (0.163203)
      kurtosis: 15.7106
      skewness: 4.07163
      entropy: 0.12719
  Image statistics:
    Overall:
      min: 0  (0)
      max: 65535 (1)
      mean: 5535.8 (0.0844709)
      standard deviation: 12318.6 (0.18797)
      kurtosis: 6.35038
      skewness: 2.60143
      entropy: 0.211481
  Total ink density: 296.471%
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Background color: cmyk(255,255,255,0)
  Border color: cmyk(223,223,223,0)
  Matte color: cmyk(189,189,189,0)
  Transparent color: cmyk(0,0,0,0)
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 1191x842+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Undefined
  Orientation: Undefined
  Profiles:
    Profile-xmp: 4551 bytes
  Properties:
    date:create: 2023-04-09T15:43:50+00:00
    date:modify: 2023-04-09T15:43:50+00:00
    date:timestamp: 2023-04-09T15:43:50+00:00
    dc:format: application/pdf
    pdf:HiResBoundingBox: 595.276x841.89+0+0
    pdf:Producer: Adobe PDF Library 16.0.3
    pdf:Trapped: False
    pdf:Version: PDF-1.6
    signature: 2a0362057debaa6ced2f424172450acc39c55b63b7820565733ccd69e31c5087
    xmp:CreateDate: 2021-12-07T12:44:46+03:00
    xmp:CreatorTool: Adobe InDesign 17.0 (Macintosh)
    xmp:MetadataDate: 2021-12-07T12:45:30+03:00
    xmp:ModifyDate: 2021-12-07T12:45:30+03:00
    xmpMM:DerivedFrom: default
    xmpMM:DocumentID: xmp.id:8f0370e4-7c3c-498d-a8d1-c2a51b6ef6dc
    xmpMM:InstanceID: uuid:e841dc2d-5e63-d842-a446-f00b5d1809bd
    xmpMM:OriginalDocumentID: adobe:docid:indd:9a729ab2-fd68-11e1-a105-e241ccc0a2a8
    xmpMM:RenditionClass: proof:pdf
  Artifacts:
    filename: PR_Nutrition_Ethiopia_PK.pdf[0]
    verbose: true
  Tainted: False
  Filesize: 3.82556MiB
  Number pixels: 1.00282M
  Pixels per second: 114.498MB
  User time: 0.020u
  Elapsed time: 0:01.008
  Version: ImageMagick 6.9.12-84 Q16 aarch64 17773 https://legacy.imagemagick.org

```

The command result looks fine to me.

`magick -colorspace sRGB -define pdf:use-cropbox=true PR_Nutrition_Ethiopia_PK.pdf[0] x.jpg`

![x](https://user-images.githubusercontent.com/13458218/230782984-40f1ae08-1356-4aa6-b2d3-1ae1628a9cd9.jpg)


**What version of Ghostscript are you using?  I am using GS 9.56.1 on IM 7.1.1.6**

**Comment by snibgo** [comment]
IM delegates the rasterisation of PDF files to Ghostscript. The image the GS returns to IM does not have an embedded profile. I don't know if GS processes embedded profiles.

Putting "`-colorspace sRGB`" before reading the PDF tells GS to make a sRGB output. 

I don't get a pink band under the maroon "project report" band.
```
C:\web\im>%IMG7%magick -verbose -colorspace sRGB -define pdf:use-cropbox=true Ethopia.pdf[0] x.png
[ghostscript library 10.00] -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 "-sDEVICE=pngalpha" -dTextAlphaBits=4 -dGraphicsAlphaBits=4 "-r72x72" -dPrinted=false -dUseCropBox -dFirstPage=1 -dLastPage=1 "-sOutputFile=C:/Users/Alan/AppData/Local/Temp/magick-eANiYLGDMltzFlB_5MoyVYXBctpXQcf8%d" "-fC:/Users/Alan/AppData/Local/Temp/magick-DRAkVweAuJ8RPENlTLSIxUwcz--_ZqHn" "-fC:/Users/Alan/AppData/Local/Temp/magick-cpfyhXwTffzY6rdoK8pGebnIcUcCF9oO"C:/Users/Alan/AppData/Local/Temp/magick-eANiYLGDMltzFlB_5MoyVYXBctpXQcf81 PNG 596x839 596x839+0+0 8-bit TrueColorAlpha sRGB 181842B 0.016u 0:00.008
Ethopia.pdf[0]=>Ethopia.pdf PDF 596x839 596x839+0+0 16-bit TrueColorAlpha sRGB 181842B 0.000u 0:00.001
writing raw profile: type=xmp, length=4551
Ethopia.pdf[0]=>x.png PDF 596x839 596x839+0+0 16-bit TrueColorAlpha sRGB 181842B 0.031u 0:00.067
```

![x](https://user-images.githubusercontent.com/33812368/230776353-11d96214-7a79-46e6-9d30-266b9b734944.png)

From the verbose text we can see my IM uses Ghostscript v10.0. What is your version? Perhaps it is old and needs upgrading.

**Reply by alanorth** [reply]
I don't know what's going on. If I use your exact command I get the following image with the pink band and incorrect gray on the bottom half:

![x](https://user-images.githubusercontent.com/191754/231937630-9a17434f-6f0f-4068-a4f0-b719b379995d.png)

Here's the verbose output:

```console
$ magick -verbose -colorspace sRGB -define pdf:use-cropbox=true data/10568-116598.pdf\[0\] /tmp/x.png
'gs' -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=pngalpha' -dTextAlphaBits=4 -dGraphicsAlphaBits=4 '-r72x72' -dPrinted=false -dUseCropBox -dFirstPage=1 -dLastPage=1 '-sOutputFile=/tmp/magick-v2GQesT9qkQ9n8ear8he4YTgoH3XPTOB%d' '-f/tmp/magick-Az3-MzX33XJRFGsK4sN3lOCQX-lw8ijl' '-f/tmp/magick-5wYANEVjTErMbo9NAKLbK5ZxuNn2MOds'
/tmp/magick-v2GQesT9qkQ9n8ear8he4YTgoH3XPTOB1 PNG 596x839 596x839+0+0 8-bit TrueColorAlpha sRGB 191256B 0.020u 0:00.008
data/10568-116598.pdf[0]=>data/10568-116598.pdf PDF 596x839 596x839+0+0 16-bit TrueColorAlpha sRGB 191256B 0.000u 0:00.000
writing raw profile: type=xmp, length=4551
data/10568-116598.pdf[0]=>/tmp/x.png PDF 596x839 596x839+0+0 16-bit TrueColorAlpha sRGB 191256B 0.130u 0:00.064
```

This is on Arch Linux with the following versions:

```console
$ pacman -Q | grep -E '(ghostscript|imagemagick)'
ghostscript 10.01.1-1
imagemagick 7.1.1.6-1
```

At this point I'm wondering if the bug is in IM, GS, or my distro's packaging one of these...

**Reply by alanorth** [reply]
> At this point I'm wondering if the bug is in IM, GS, or my distro's packaging one of these...

I just checked this same PDF in [ImageMagick 7 via a Docker environment](https://hub.docker.com/r/dpokidov/imagemagick) and the results were correct (at least consistent with what I see in the PDF). That Docker environment uses Ghostscript 9.53.3, so I am wondering if it's a Ghostscript issue, either in general or with Arch's packaging.

**Comment by urban-warrior** [comment]
Learn how ImageMagick version 7 differs from previous versions with our [porting guide](https://imagemagick.org/script/porting.php).

**Reply by alanorth** [reply]
I'm using ImageMagick via im4java in one project and all my servers still have ImageMagick 6.x for now, though now I'm wondering how some of our code will break when ImageMagick 7.x lands on those servers. Also, I see that im4java hasn't been updated since 2012! Ouch...
